### ISSUE TITLE
Inconsistency between nested @Transaction and inTransaction(..)

### ISSUE THREAD

**Issue by sarxos** [jdbi/jdbi#1350]
Hello @jdbi team,

I came across some inconsistent mechanism. Documentation for `@Transaction` states:

>  Nested <code>@Transaction</code> annotations (e.g. one method calls another method, where both methods have this annotation) are collapsed into a single transaction. If the outer method annotation specifies an isolation level, then the inner method must either specify the same level, or not specify any level.

Which is very nice since it would be extremely hard to track transaction without having `Handle` available in transactional method scope.

But when I tried transaction nesting with SQL objects:

```java
final int id = dao(TransactionalInterfaceDao.class).inTransaction(dao1 -> {
  return dao(TransactionalInterfaceDao.class).inTransaction(dao2 -> {
    return dao2.get(dao2.insert(msg.getValue()));
  });
});
```

Such as:

```java
interface TransactionalInterfaceDao extends Transactional<TransactionalInterfaceDao> {

  @GetGeneratedKeys
  @SqlUpdate("insert into test (value) values (:value)")
  int insert(@Bind("value") int value);

  @SqlQuery("select value from test where id = :id")
  int get(@Bind("id") int id);
}
```

I got:

```plain
Caused by: java.lang.IllegalStateException: Already in transaction
	at org.jdbi.v3.core.transaction.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:129)
	at org.jdbi.v3.core.Handle.inTransaction(Handle.java:393)
	at org.jdbi.v3.sqlobject.transaction.Transactional.inTransaction(Transactional.java:96)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.jdbi.v3.sqlobject.DefaultMethodHandler.invoke(DefaultMethodHandler.java:102)
	at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$null$13(SqlObjectFactory.java:163)
	at org.jdbi.v3.core.ConstantHandleSupplier.invokeInContext(ConstantHandleSupplier.java:52)
	at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$createInvocationHandler$14(SqlObjectFactory.java:162)
	at com.sun.proxy.$Proxy51.inTransaction(Unknown Source)
	***
```

I was under impression that declarative approach given by `@Transaction` annotation and imperative approach given by `Transactional` interface should work very similar, but they are very different when it comes to nesting.

I noticed #937, which is a nice approach to translate nested transaction as a sub-transaction, but this is applicable only in case when user is working on a RDBMS which supports sub-transactions specifically.

Right now I'm in the middle of changing `inTransaction()` to `@Transaction` because there is no need to track nested transactions, but I wanted to point this out and suggest that maybe, just maybe, both approaches should handle transactions nesting the same way. Shouldn't default behaviour for this be the same for both mechanisms? Use not-yet-implemented sub-transactions if they are supported by RDBMS, or collapse transactions to the uppermost one otherwise?

**Comment by qualidafial** [comment]
I would be comfortable with this change, provided we enforce the same semantics for transaction isolation. @stevenschlansker, @TheRealMarnes ?

**Comment by qualidafial** [comment]
@stevenschlansker @TheRealMarnes any input on this ticket, before I start work?

**Comment by stevenschlansker** [comment]
I think that makes sense.  Hopefully the change is that you move the SQLObject implementation to core and just call it from there.

**Comment by qualidafial** [comment]
That's exactly my plan

**Comment by qualidafial** [comment]
This has been fixed as of 3.7.0, now in Maven central.
